pkg/controller/identity/iamuserpolicyattachment/controller.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	userClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: userClient, kube: c.kube}, errors.Wrap(err, errCreateUserClient)
+}
+
+type external struct {
+	client iam.UserPolicyAttachmentClient
+	kube   client.Client
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMUserPolicyAttachment)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	observed, err := e.client.ListAttachedUserPoliciesRequest(&awsiam.ListAttachedUserPoliciesInput{
+		UserName: cr.Spec.ForProvider.UserName,
+	}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(iam.IsErrorNotFound, err), errGet)
+	}
+
+	var attachedPolicyObject *awsiam.AttachedPolicy
+	for i, policy := range observed.AttachedPolicies {
+		if cr.Spec.ForProvider.PolicyARN == aws.StringValue(policy.PolicyArn) {
+			attachedPolicyObject = &observed.AttachedPolicies[i]
+			break
+		}
+	}
+
+	if attachedPolicyObject == nil {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	iam.LateInitializeUserPolicy(&cr.Spec.ForProvider, attachedPolicyObject)
+	if !cmp.Equal(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+
+	cr.SetConditions(runtimev1alpha1.Available())
+
+	cr.Status.AtProvider = v1alpha1.IAMUserPolicyAttachmentObservation{
+		AttachedPolicyARN: aws.StringValue(attachedPolicyObject.PolicyArn),
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: true,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMUserPolicyAttachment)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.SetConditions(runtimev1alpha1.Creating())
+
+	_, err := e.client.AttachUserPolicyRequest(&awsiam.AttachUserPolicyInput{
+		PolicyArn: aws.String(cr.Spec.ForProvider.PolicyARN),
+		UserName:  cr.Spec.ForProvider.UserName,
+	}).Send(ctx)
+
+	return managed.ExternalCreation{}, errors.Wrap(err, errAttach)
+}
+
+func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
+	// PolicyARN is the only distinguishing field and on update to that, new policy is attached
```

Comment:
I see that the `PolicyARN` field is marked immutable. Are we saying that we don't support updating an `IAMUserPolicyAttachment`, and that instead the user will create a new `IAMUserPolicyAttachment` and delete the old one?

If so that seems like a good approach to me, but it could be worth updating this comment. At first I thought we were saying we'd allow the user to update the `PolicyARN` field and would call `Create` again to attach the new policy (thus orphaning the old one).

Reply:
Updated the comment, @negz